### PR TITLE
feat: supporting esm import

### PR DESCRIPTION
**Pull request by anuragmerndev**
add the exports in package.json to support esm import for js and ts

## Summary of change

This pull request addresses issue [#783](https://github.com/supertokens/supertokens-node/issues/783) by adding [subpath patterns](https://nodejs.org/api/packages.html#subpath-patterns) exports to the package.json file. These exports support ESM (ECMAScript Modules) imports for both JavaScript and TypeScript files, improving compatibility and ease of use for developers. Now, imports like import Session from "supertokens-node/recipe/session"; can be used without specifying the index.js file explicitly.

## Related issues

-   [Link to issue #783](https://github.com/supertokens/supertokens-node/issues/783)
-   [Link to issue #459](https://github.com/supertokens/supertokens-node/issues/459)

## Test Plan

1. Update the package.json file with the provided exports configuration.
2. Ensure the project builds without errors.
3. Verify that existing imports using ESM syntax work as expected.
4. Test importing modules in both JavaScript and TypeScript files.
5. Run unit tests to confirm that the changes do not introduce regressions.
6. Check for any compatibility issues with existing codebases using the library.

https://github.com/supertokens/supertokens-node/assets/144275260/197d89a4-2817-4a33-9334-85adccfba59f

## Documentation changes

No documentation changes are required for this PR.

## Checklist for important updates

-   [ ] Changelog has been updated
-   [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
    -   Along with the associated array in `lib/ts/version.ts`
-   [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
-   [ ] Changes to the version if needed
    -   In `package.json`
    -   In `package-lock.json`
    -   In `lib/ts/version.ts`
-   [ ] Had run `npm run build-pretty`
-   [ ] Had installed and ran the pre-commit hook
-   [ ] If new thirdparty provider is added,
    -   [ ] update switch statement in `recipe/thirdparty/providers/configUtils.ts` file, `createProvider` function.
    -   [ ] add an icon on the user management dashboard.
-   [ ] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.
-   [ ] If have added a new web framework, update the `add-ts-no-check.js` file to include that
-   [ ] If added a new recipe / api interface, then make sure that the implementation of it uses NON arrow functions only (like `someFunc: function () {..}`).
-   [ ] If added a new recipe, then make sure to expose it inside the recipe folder present in the root of this repo. We also need to expose its types.

## Remaining TODOs for this PR

No additional tasks are required for this PR.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -27,6 +27,12 @@
         "type": "git",
         "url": "git+https://github.com/supertokens/supertokens-node.git"
     },
+    "exports": {
+        "./recipe/*": {
+            "types": "./recipe/*/index.d.ts",
+            "default": "./recipe/*/index.js"
```
Comment (ITenthusiasm) [inline package.json:33]: Were you able to test this with both ESM and CJS node applications? I'd be surprised if we didn't need distinct export paths for CJS/ESM

**Comment by ITenthusiasm** [comment]
From the video, I can see that the build works; but I don't see the examples of ESM-based/CJS-based/TypeScript Node applications working properly. How did those go?

**Comment by ITenthusiasm** [comment]
Also just giving a reminder to target the proper version branch, as mentioned in the PR checklist.

File: package.json (hunk 1 above)
Comment (anuragmerndev) [reply package.json:33]: Yeah in the test video you can see in the test-js side I have tried importing while having type module and without it as well 
Using require and import both

**Comment by anuragmerndev** [comment]
> From the video, I can see that the build works; but I don't see the examples of ESM-based/CJS-based/TypeScript Node applications working properly. How did those go?

I have updated the video last video didn't capture the other two windows
attaching here as well 


https://github.com/supertokens/supertokens-node/assets/144275260/900d6be1-a269-4da5-949c-ff134016bfab

**Comment by anuragmerndev** [comment]
> Also just giving a reminder to target the proper version branch, as mentioned in the PR checklist.

Okay I have updated to the latest unreleased branch

**Comment by ITenthusiasm** [comment]
Thanks for the update! As a heads up, I probably won't be able to look at this until the weekend. There's also `rishabhpoddar` when he has more availability, though. I'm mainly here to help with making sure that the import/export configuration looks normal.

An optional recommendation: Your commit history might be clearer if you `rebase` your current changes onto the latest version of the target branch instead of merging that branch's features into your own.

**Comment by ITenthusiasm** [comment]
So var the new video looks promising though!

**Comment by anuragmerndev** [comment]
> Thanks for the update! As a heads up, I probably won't be able to look at this until the weekend. There's also `rishabhpoddar` when he has more availability, though. I'm mainly here to help with making sure that the import/export configuration looks normal.
> 
> An optional recommendation: Your commit history might be clearer if you `rebase` your current changes onto the latest version of the target branch instead of merging that branch's features into your own.

Sure I'll do the rebase

**Comment by ITenthusiasm** [comment]
Hey friend, sorry I wasn't able to get to this. The weekend actually turned out much busier than I expected. But it looks like things should calm down after today so hopefully I'll be able to review soon.

(I know the code change is small. But I want to be able to test the changes as well to double-verify that no unsuspecting breaking changes would sneak in. I'm still a little surprised that we wouldn't need separate files for handling CJS/ESM exports.)

**Comment by ITenthusiasm** [comment]
@anuragmerndev Can you link this to #459 as well to help with cleaning up the issues?

---

Separately, I'm probably going to try to pull this branch locally to test (just to doubly make sure). My first suspicion as to why this would work is that ESM is able to import from CJS (but not the other way around). If that's the case then yeah, this should be smooth sailing.

**Review by ITenthusiasm — APPROVED** [review]
LGTM (Sorry this took so long to review. Been really busy.)

## @rishabhpoddar 
I only tested importing from `supertokens-node/recipe/emailpassword`. (@anuragmerndev's example uses significantly more `import`s.) But for that import, I can confirm that `require`ing worked just fine, and `import`ing worked for `type=module` node apps. (`import`ing does not work if the semantic `exports` property is excluded from the `package.json` file in `supertokens-node`).

### Some Brief Background
CJS applications cannot import from ESM applications. But ESM applications can import from _both_ CJS and ESM applications. The `supertokens-node` package does not identify as an ESM package (because `type="module"` is not in the `package.json` file); so it is a CJS package. Therefore, when an ESM application tries to import from `supertokens-node`, it sees it as a CJS package, and the import works fine (because ESM can handle those). Of course, regular CJS applications have no problem importing from `supertokens-node` currently.

### Additional Notes
_If there are any other paths that people should be allowed to `import`/`require` from, they need to be added to `exports` in `package.json`. Otherwise, consumers won't be allowed to access those paths._ The `exports` property in `package.json` is effectively a "whitelist". If it's sufficient to only allow `import`s from `supertokens-node/recipe/**/*`, then I guess this PR is good to go. But if other paths need to be exposed, it would be helpful to know those.

Not sure what you wanted to do about tests. There's probably a way to write those? The test would basically involve successfully `import`ing/`require`ing without errors -- if you wanted to automate that.

If you guys added documentation on how to write imports for applications using ESM, then that documentation will need to be re-updated when this gets merged. (Otherwise, don't worry about it.)

## @anuragmerndev
Thanks for looking into this. (And sorry again for the delay.) I'm sure this change will make a lot of people happy. Could you share a code snippet of the test file that you used in the video? That way, if anyone else wants to test locally, they'll have the ability to do so with ease.

**Comment by anuragmerndev** [comment]
> LGTM (Sorry this took so long to review. Been really busy.)
> 
> ## @rishabhpoddar
> I only tested importing from `supertokens-node/recipe/emailpassword`. (@anuragmerndev's example uses significantly more `import`s.) But for that import, I can confirm that `require`ing worked just fine, and `import`ing worked for `type=module` node apps. (`import`ing does not work if the semantic `exports` property is excluded from the `package.json` file in `supertokens-node`).
> 
> ### Some Brief Background
> CJS applications cannot import from ESM applications. But ESM applications can import from _both_ CJS and ESM applications. The `supertokens-node` package does not identify as an ESM package (because `type="module"` is not in the `package.json` file); so it is a CJS package. Therefore, when an ESM application tries to import from `supertokens-node`, it sees it as a CJS package, and the import works fine (because ESM can handle those). Of course, regular CJS applications have no problem importing from `supertokens-node` currently.
> 
> ### Additional Notes
> _If there are any other paths that people should be allowed to `import`/`require` from, they need to be added to `exports` in `package.json`. Otherwise, consumers won't be allowed to access those paths._ The `exports` property in `package.json` is effectively a "whitelist". If it's sufficient to only allow `import`s from `supertokens-node/recipe/**/*`, then I guess this PR is good to go. But if other paths need to be exposed, it would be helpful to know those.
> 
> Not sure what you wanted to do about tests. There's probably a way to write those? The test would basically involve successfully `import`ing/`require`ing without errors -- if you wanted to automate that.
> 
> If you guys added documentation on how to write imports for applications using ESM, then that documentation will need to be re-updated when this gets merged. (Otherwise, don't worry about it.)
> 
> ## @anuragmerndev
> Thanks for looking into this. (And sorry again for the delay.) I'm sure this change will make a lot of people happy. Could you share a code snippet of the test file that you used in the video? That way, if anyone else wants to test locally, they'll have the ability to do so with ease.


@ITenthusiasm  Thank you for the update on the issue, I have created a repo and added readme for the test for this fix 
[repo-to-test-imports](https://github.com/anuragmerndev/supertokens-feat-783)

And yeah thank you @ITenthusiasm for sharing your time on checking this PR, means a lot. 

Also @rishabhpoddar, @ITenthusiasm is correct for the exports parts as I have only exported from recipe directory no other files would be accessible apart from those in recipe (it's a whitelist), so do let me know if we need to export other files as well as when I was going through the documentation for usage of **supertokens-node** I was not able to find any imports other than recipes. 

If we need to add any other imports I'll update the PR with the same.

**Comment by rishabhpoddar** [comment]
Thanks @anuragmerndev and @ITenthusiasm for the work so far, really appreciate it. Someone from our team will have a look at this too soon

**Review by porcellus — CHANGES_REQUESTED** [review]
Hi,
First of all, sorry for the late review. We've been swamped releasing 18.0 and other projects.

As for the PR, I have some concerns:
- Please update the PR (and the target branch) to the latest
- Some entry points are missing (see my comment), so ideally, we'd like to figure out a way to "test" if this happens again. This is not necessarily a blocker for this PR, but any ideas are welcome. 
- If we can test that the defined exports match some files, that'd also be awesome.
- We need to allow importing from the build output directly, as we've seen some use cases that require that. I know it's less than ideal, but we want to enable pretty extreme customizations. I was thinking of adding something like the following to the exports obj, but it seems relatively awkward, so any ideas are welcome:
```
        "./lib/*.js": {
            "types": "./lib/*.d.ts",
            "default": "./lib/*.js"
        },
        "./lib/*": {
            "types": "./lib/*.d.ts",
            "default": "./lib/*.js"
        }
```
- Can we also support existing apps that import us using a ".js" module specifier? Otherwise, this PR will be a breaking change.

**Comment by anuragmerndev** [comment]
> Hi, First of all, sorry for the late review. We've been swamped releasing 18.0 and other projects.
> 
> As for the PR, I have some concerns:
> 
> * Please update the PR (and the target branch) to the latest
> * Some entry points are missing (see my comment), so ideally, we'd like to figure out a way to "test" if this happens again. This is not necessarily a blocker for this PR, but any ideas are welcome.
> * If we can test that the defined exports match some files, that'd also be awesome.
> * We need to allow importing from the build output directly, as we've seen some use cases that require that. I know it's less than ideal, but we want to enable pretty extreme customizations. I was thinking of adding something like the following to the exports obj, but it seems relatively awkward, so any ideas are welcome:
> 
> ```
>         "./lib/*.js": {
>             "types": "./lib/*.d.ts",
>             "default": "./lib/*.js"
>         },
>         "./lib/*": {
>             "types": "./lib/*.d.ts",
>             "default": "./lib/*.js"
>         }
> ```
> 
> * Can we also support existing apps that import us using a ".js" module specifier? Otherwise, this PR will be a breaking change.

Hello @porcellus 

No worries about the late reviews, In fact, thank you for checking the PR also here's the updates on your concerns

- I have updated the PR target branch to 18.0.
- I'll add the missing entry points in the final commit I am working on.
- I can provide the repository which I shared before in the comments for testing.
- For the build exports, I am researching to find a better solution; otherwise, we can proceed with the approach you provided.
- We will need to add something like the following to support existing apps using the ".js" module specifier, so this PR won't be a breaking change:

```
  "./recipe/*": {
        "types": "./recipe/*/index.d.ts",
        "default": "./recipe/*/index.js"
    },
    // for index.js support
    "./recipe/*/index.js": {
        "types": "./recipe/*/index.d.ts",
        "default": "./recipe/*/index.js"
    }
```

**Comment by porcellus** [comment]
> I can provide the repository which I shared before in the comments for testing.

I meant some way of doing this in an automated test, especially testing that we didn't miss anything.

> For the build exports, I am researching to find a better solution; otherwise, we can proceed with the approach you provided.

That'd be awesome.

> We will need to add something like the following to support existing apps using the ".js" module specifier...

Sounds good

> Also I wanted to confirm that do we need to provide this .js support for all the exports you mentioned in the comments?

Yeah, all entrypoints should work either way, otherwise we might break some apps.

**Comment by anuragmerndev** [comment]
> > I can provide the repository which I shared before in the comments for testing.
> 
> I meant some way of doing this in an automated test, especially testing that we didn't miss anything.
> 
> > For the build exports, I am researching to find a better solution; otherwise, we can proceed with the approach you provided.
> 
> That'd be awesome.
> 
> > We will need to add something like the following to support existing apps using the ".js" module specifier...
> 
> Sounds good
> 
> > Also I wanted to confirm that do we need to provide this .js support for all the exports you mentioned in the comments?
> 
> Yeah, all entrypoints should work either way, otherwise we might break some apps.



> > I can provide the repository which I shared before in the comments for testing.
> 
> I meant some way of doing this in an automated test, especially testing that we didn't miss anything.
> 
> > For the build exports, I am researching to find a better solution; otherwise, we can proceed with the approach you provided.
> 
> That'd be awesome.
> 
> > We will need to add something like the following to support existing apps using the ".js" module specifier...
> 
> Sounds good
> 
> > Also I wanted to confirm that do we need to provide this .js support for all the exports you mentioned in the comments?
> 
> Yeah, all entrypoints should work either way, otherwise we might break some apps.

Hello @porcellus,

I have pushed the updated code with support for rest of modules and the .js modules for existing apps along with build folder.

About the test thing I am still researching how can we automate tests for this specific PR, haven't found any till now.

Kindly review the same and let me know for any updates.

**Review by porcellus — APPROVED** [review]

**Comment by porcellus** [comment]
I've reviewed this and it looks good. I'll merge this into an intermediary branch, because I need to add some further changes (I've figured out a way to automatically test it) and add a changelog entry/increase the version number.

**Comment by anuragmerndev** [comment]
> I've reviewed this and it looks good. I'll merge this into an intermediary branch, because I need to add some further changes (I've figured out a way to automatically test it) and add a changelog entry/increase the version number.

Alright thank you,

Also can you please share the logic for automated tests

**Comment by porcellus** [comment]
You can check out the test [here](https://github.com/supertokens/supertokens-node/pull/850/files#diff-e637e60527f33e028c60460a98d3c7957a94e752f32b57e1d94b7c54c56e5e38). It's far from pretty, but it already showed some errors we had.

Also, if you'd like, please feel free to add yourself as a contributor in the supertokens-core repo. (similar to [this PR](https://github.com/supertokens/supertokens-core/pull/996))
Thank you for contributing to SuperTokens.

**Comment by anuragmerndev** [comment]
> You can check out the test [here](https://github.com/supertokens/supertokens-node/pull/850/files#diff-e637e60527f33e028c60460a98d3c7957a94e752f32b57e1d94b7c54c56e5e38). It's far from pretty, but it already showed some errors we had.
> 
> Also, if you'd like, please feel free to add yourself as a contributor in the supertokens-core repo. (similar to [this PR](https://github.com/supertokens/supertokens-core/pull/996)) Thank you for contributing to SuperTokens.

Thank you for letting me contribute will do alot more of it
Also here's the PR to the Contributor [update](https://github.com/supertokens/supertokens-core/pull/1003) Do let me know if need any update

**Pull request merged**